packages-exp/auth-exp/src/core/auth/auth_impl.test.ts

```diff
@@ -117,6 +118,138 @@ describe('AuthImpl', () => {
       );
     });
   });
+
+  describe('change listeners', () => {
+    // // Helpers to convert auth state change results to promise
+    // function onAuthStateChange(callback: NextFn<User|null>)
+
+    it('immediately calls authStateChange if initialization finished', done => {
+      const user = testUser('uid');
+      auth.currentUser = user;
+      auth._isInitialized = true;
+      auth.onAuthStateChanged(user => {
+        expect(user).to.eq(user);
+        done();
+      });
+    });
+
+    it('immediately calls idTokenChange if initialization finished', done => {
+      const user = testUser('uid');
+      auth.currentUser = user;
+      auth._isInitialized = true;
+      auth.onIdTokenChange(user => {
+        expect(user).to.eq(user);
+        done();
+      });
+    });
+
+    it('immediate callback is done async', () => {
+      auth._isInitialized = true;
+      let callbackCalled = false;
+      auth.onIdTokenChange(() => {
+        callbackCalled = true;
+      });
+
+      expect(callbackCalled).to.be.false;
+    });
+
+    describe('user logs in/out, tokens refresh', () => {
+      let user: User;
+      let callback: sinon.SinonSpy;
+
+      beforeEach(() => {
+        user = testUser('uid');
+        callback = sinon.spy();
+      });
+
+      it('onAuthStateChange triggers on log in', async () => {
+        auth.onAuthStateChanged(callback);
```

Comment:
these first three lines of setup are re-used a ton, can you extract to a context() block with a beforeEeach()?

Reply:
Done